src/java.base/share/classes/java/io/ObjectStreamReflection.java

```diff
@@ -0,0 +1,173 @@
+     * @throws IOException if the call to {@link ObjectInputStream#readFields} or one of its field accessors throws this exception type
+     */
+    private static void defaultWriteObject(ObjectStreamClass streamClass, Object obj, ObjectOutputStream oos) throws IOException {
+        ObjectOutputStream.PutField putField = oos.putFields();
+        byte[] bytes = new byte[streamClass.getPrimDataSize()];
+        Object[] objs = new Object[streamClass.getNumObjFields()];
+        streamClass.getPrimFieldValues(obj, bytes);
+        streamClass.getObjFieldValues(obj, objs);
+        for (ObjectStreamField field : streamClass.getFields(false)) {
+            int offset = field.getOffset();
+            String fieldName = field.getName();
+            switch (field.getTypeCode()) {
+                case 'B' -> putField.put(fieldName, bytes[offset]);
+                case 'C' -> putField.put(fieldName, ByteArray.getChar(bytes, offset));
+                case 'D' -> putField.put(fieldName, ByteArray.getDouble(bytes, offset));
+                case 'F' -> putField.put(fieldName, ByteArray.getFloat(bytes, offset));
+                case 'I' -> putField.put(fieldName, ByteArray.getInt(bytes, offset));
+                case 'J' -> putField.put(fieldName, ByteArray.getLong(bytes, offset));
+                case 'S' -> putField.put(fieldName, ByteArray.getShort(bytes, offset));
+                case 'Z' -> putField.put(fieldName, ByteArray.getBoolean(bytes, offset));
+                case '[', 'L' -> putField.put(fieldName, objs[offset]);
+                default -> throw new IllegalStateException();
+            }
+        }
+        oos.writeFields();
+    }
+
+    static final class Access implements JavaObjectStreamReflectionAccess {
+        static {
+            SharedSecrets.setJavaObjectStreamReflectionAccess(new Access());
+        }
+
+        public MethodHandle defaultReadObject(Class<?> clazz) {
+            ObjectStreamClass streamClass = getStreamClass(clazz);
+            return streamClass == null ? null : DRO_HANDLE.bindTo(streamClass).asType(MethodType.methodType(void.class, streamClass.forClass(), ObjectInputStream.class));
+        }
+
+        public MethodHandle defaultWriteObject(Class<?> clazz) {
+            ObjectStreamClass streamClass = getStreamClass(clazz);
+            return streamClass == null ? null : DWO_HANDLE.bindTo(streamClass).asType(MethodType.methodType(void.class, streamClass.forClass(), ObjectOutputStream.class));
+        }
+
+        private static ObjectStreamClass getStreamClass(final Class<?> clazz) {
+            ObjectStreamClass streamClass = ObjectStreamClass.lookup(clazz);
+            if (streamClass == null) {
+                return null;
+            }
+            try {
+                streamClass.checkDefaultSerialize();
+            } catch (InvalidClassException e) {
+                return null;
+            }
+            return streamClass;
+        }
```

Comment:
Possibly move common setup into the helper function:

```suggestion
        public MethodHandle defaultReadObject(Class<?> clazz) {
            return handleForClass(DRO_HANDLE, clazz, ObjectInputStream.class);
        }

        public MethodHandle defaultWriteObject(Class<?> clazz) {
            return handleForClass(DWO_HANDLE, clazz, ObjectOutputStream.class);
        }

        private static MethodHandle handleForClass(MethodHandle handle, final Class<?> clazz, Class<?> ioClass) {
            ObjectStreamClass streamClass = ObjectStreamClass.lookup(clazz);
            if (streamClass != null) {
                try {
                    streamClass.checkDefaultSerialize();
                    return handle.bindTo(streamClass)
                            .asType(MethodType.methodType(void.class, clazz, ioClass));
                } catch (InvalidClassException e) {
                    // ignore and return null
                }
            }
            return null;
        }

```